### PR TITLE
Add support for GIT model in VQA pipelines

### PR DESCRIPTION
**Pull request by marechaux**
# What does this PR do?

This PR implement support for generative model in VQA pipeline (and more precisely GIT model).

Fixes part of #21110

This is my first contribution here, I am uncertain if my approach is correct. Please advise me if any modifications are necessary 😃 

## Before submitting
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#start-contributing-pull-requests),
      Pull Request section?
- [x] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case. => #21110 
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

@NielsRogge

### CODE REVIEW

**Comment by marechaux** [comment]
I have two remarks here : 
* The CI is red due to this change : https://github.com/huggingface/transformers/pull/23348/files#diff-b452cc4f093e4b991e92054cf7d504edab44be07c4957969df85a5562238313cR48 the model now used in test is `hf-internal-testing/tiny-random-ViltForQuestionAnswering` instead of `hf-internal-testing/tiny-vilt-random-vqa` (and it seems like the new model image processor can't process images, I am not sure about how to fix it)
* The test for GIT in VQA pipeline doesn't run because `hf-internal-testing/tiny-random-GitForVisualQuestionAnswering` doesn't exist, I need help about this point as well

File: docs/source/en/model_doc/git.mdx [hunk 1]
```diff
@@ -73,4 +73,9 @@ The resource should ideally demonstrate something new instead of duplicating an
 ## GitForCausalLM
 
 [[autodoc]] GitForCausalLM
-    - forward
\ No newline at end of file
+    - forward
+
+## GitForVisualQuestionAnswering
```
Comment (NielsRogge) [inline docs/source/en/model_doc/git.mdx:78]: Hi @marechaux thanks a lot for working on this!

However we don't need to add a new class for it actually, as GitForCausalLM handles VQA as shown in t[his code snippet](https://huggingface.co/docs/transformers/model_doc/git#transformers.GitForCausalLM.forward.example-2)

Comment (marechaux) [reply docs/source/en/model_doc/git.mdx]: Thanks, I just rewrite it without adding a new model

File: src/transformers/pipelines/visual_question_answering.py [hunk 2]
```diff
@@ -118,20 +136,55 @@ def preprocess(self, inputs, padding=False, truncation=False):
         model_inputs.update(image_features)
         return model_inputs
 
-    def _forward(self, model_inputs):
-        model_outputs = self.model(**model_inputs)
+    def _forward(self, model_inputs, top_k=5, max_new_tokens=None, num_beams=None):
```
Comment (NielsRogge) [inline src/transformers/pipelines/visual_question_answering.py:139]:
```suggestion
    def _forward(self, model_inputs, **generate_kwargs):
```
I think the forward should just accept a `generate_kwargs` like the [text generation pipeline](https://github.com/huggingface/transformers/blob/3658488ff77ff8d45101293e749263acf437f4d5/src/transformers/pipelines/text_generation.py#LL231C38-L231C57)

File: src/transformers/pipelines/visual_question_answering.py [hunk 3]
```diff
@@ -15,6 +15,11 @@
 logger = logging.get_logger(__name__)
 
 
+class ModelType(ExplicitEnum):
+    CLASSIFIER = "classifier"
+    GIT = "GIT"
```
Comment (NielsRogge) [inline src/transformers/pipelines/visual_question_answering.py:20]: Maybe we can make a distinction between classifier vs generative. GIT is a generative model

File: src/transformers/pipelines/visual_question_answering.py [hunk 4]
```diff
@@ -55,15 +60,28 @@ def __init__(self, *args, **kwargs):
...
         self.check_model_type(MODEL_FOR_VISUAL_QUESTION_ANSWERING_MAPPING)
 
-    def _sanitize_parameters(self, top_k=None, padding=None, truncation=None, **kwargs):
-        preprocess_params, postprocess_params = {}, {}
+        if self.model.config.__class__.__name__ == "GitConfig":
+            self.model_type = ModelType.GIT
+        elif self.model.config.__class__.__name__ in ["ViltConfig"]:
+            self.model_type = ModelType.CLASSIFIER
+
+    def _sanitize_parameters(
```
Comment (NielsRogge) [inline src/transformers/pipelines/visual_question_answering.py:68]: The `_sanitize_parameters` method can look a lot like the one of the text generation pipeline: https://github.com/huggingface/transformers/blob/3658488ff77ff8d45101293e749263acf437f4d5/src/transformers/pipelines/text_generation.py#L90. No need to include things like `num_beams` here

**Comment by marechaux** [comment]
Thanks for the review, I updated my changes following your comments. However, I have several doubt on my approach. 

### Beam search for scores 
I use beam scores to provide a score to follow the "signature" of the pipeline described [here](https://github.com/huggingface/transformers/blob/c3c9b03d55f2d8094a2ac058db566d469baa8bbd/src/transformers/pipelines/visual_question_answering.py#L106-L110). Is it a correct ? 

The beam search is so slow that it makes the pipeline test [timeout](https://app.circleci.com/pipelines/github/huggingface/transformers/65019/workflows/be9d4fde-5cbb-4db0-ba85-0110d8988953/jobs/806494), it runs locally but in more than 120s,  which make me think I'm not in the right way here

### Tokenizer padding
Also when I use the pipeline with `microsoft/git-base-textvqa` in batch mode, I have this warning : 
```
A decoder-only architecture is being used, but right-padding was detected! For correct generation results, please set `padding_side='left'` when initializing the tokenizer.
```
This warning is legitimate, but I don't know how to fix it as this parameter can only be set at tokenizer init. 

### Broken `Vilt` tests
Due to [this change](https://github.com/huggingface/transformers/pull/23348/files#diff-b452cc4f093e4b991e92054cf7d504edab44be07c4957969df85a5562238313cR48) , the model used in unit test for Vilt model is now `hf-internal-testing/tiny-random-ViltForQuestionAnswering` instead of `hf-internal-testing/tiny-vilt-random-vqa`.

It seems like the new model image processor can't process images, I am not sure about how to fix it

Should I fix the model directly in the hub ?

**Comment by marechaux** [comment]
Gently ping here @NielsRogge  :-)

File: src/transformers/pipelines/visual_question_answering.py [hunk 5]
```diff
@@ -118,20 +131,56 @@ def preprocess(self, inputs, padding=False, truncation=False):
...
+            if "max_new_tokens" not in generate_kwargs:
+                # defaulting max_new_tokens to 100
+                generate_kwargs["max_new_tokens"] = 100
+            generate_outputs = self.model.generate(**model_inputs, **generate_kwargs)
+            model_outputs = {
+                "sequences_scores": generate_outputs.sequences_scores.reshape(
+                    (model_inputs["input_ids"].shape[0], top_k)
+                ),
+                "sequences": generate_outputs.sequences.reshape((model_inputs["input_ids"].shape[0], top_k, -1)),
+            }
```
Comment (NielsRogge) [inline src/transformers/pipelines/visual_question_answering.py:157]: No need to all of this, I'd just run a `generate` as done [here](https://github.com/huggingface/transformers/blob/539e2281cd97c35ef4122757f26c88f44115fa94/src/transformers/pipelines/text_generation.py#L263). No need to check things like `num_beams`, `num_return_sequences`, `output_scores`, etc.

For a generative model, it will just generate an answer so no need for any top k scores. For a classifier, the top k scores make sense as we can for instance take the top 5 predicted class

File: src/transformers/pipelines/visual_question_answering.py [hunk 6]
```diff
@@ -118,20 +131,56 @@ def preprocess(self, inputs, padding=False, truncation=False):
...
-            scores, ids = probs.topk(top_k)
+            if self.model_type == ModelType.CLASSIFIER:
+                probs = model_outputs.logits.sigmoid()[0]
+                scores, ids = probs.topk(top_k)
+                ids = ids.tolist()
+                answers = [self.model.config.id2label[_id] for _id in ids]
+            elif self.model_type == ModelType.GENERATIVE:
+                scores = model_outputs["sequences_scores"][0]
+                decoded_outputs = self.tokenizer.batch_decode(model_outputs["sequences"][0], skip_special_tokens=False)
+                answers = [self.postprocess_git_output_single(o) for o in decoded_outputs]
```
Comment (NielsRogge) [inline src/transformers/pipelines/visual_question_answering.py:174]: This looks OK

File: src/transformers/pipelines/visual_question_answering.py (hunk 5 above)
Comment (NielsRogge) [reply src/transformers/pipelines/visual_question_answering.py:157]: Also the forward here should only add `**generate_kwargs`
